File: packages/ruby/lib/html_to_markdown.rb
# frozen_string_literal: true

require_relative 'html_to_markdown/version'
require 'html_to_markdown_rb'

module HtmlToMarkdown
  autoload :CLI, 'html_to_markdown/cli'
  autoload :CLIProxy, 'html_to_markdown/cli_proxy'

  class << self
    alias native_convert convert
  end

  module_function

  # Convert HTML to Markdown, returning a Hash with:
  #   - :content [String, nil] the converted Markdown output
  #   - :document [nil] document structure (not yet exposed)
  #   - :metadata [Hash, nil] extracted HTML metadata
  #   - :tables [Array<Hash>] extracted tables with :grid and :markdown
  #   - :images [Array<Hash>] extracted inline images
  #   - :warnings [Array<Hash>] processing warnings
  #
  # @param html [String] HTML string to convert
  # @param options [Hash, nil] optional conversion options
  # @return [Hash] conversion result
  def convert(html, options = nil)
    native_convert(html.to_s, options)
  end
end

File: packages/elixir/lib/html_to_markdown.ex
defmodule HtmlToMarkdown do
  @moduledoc """
  High-level Elixir interface for the Rust html-to-markdown engine.
  """

  alias HtmlToMarkdown.{
    Error,
    Native,
    Options
  }

  @type options_input :: Options.t() | map() | keyword() | nil

  @doc """
  Convert HTML to Markdown, returning a `ConversionResult` map with:
  - `:content` - the converted Markdown string (or nil in extraction-only mode)
  - `:metadata` - extracted HTML metadata (document, headers, links, images, structured_data)
  - `:tables` - list of extracted tables with `:grid` (rows, cols, cells) and `:markdown`
  - `:warnings` - list of processing warnings with `:message` and `:kind`

  Returns `{:ok, result}` or `{:error, reason}`.
  """
  @spec convert(String.t(), options_input()) :: {:ok, map()} | {:error, term()}
  def convert(html, options \\ nil) when is_binary(html) do
    options_map = normalize_options(options) || %{}

    Native.convert(html, options_map)
  end

  @doc """
  Bang variant of `convert/2`. Raises on failure.
  """
  @spec convert!(String.t(), options_input()) :: map()
  def convert!(html, options \\ nil) do
    case convert(html, options) do
      {:ok, result} -> result
      {:error, reason} -> raise Error, message: inspect(reason)
    end
  end

  defp normalize_options(nil), do: nil

  defp normalize_options(options) do
    options
    |> Options.new()
    |> Options.to_map()
  end
end
